src/common/Language.ts

```diff
@@ -64,7 +89,11 @@ export interface IdentifierGenerator {
   general: (i: number) => string;
 }
 
-export type Emitter = (program: IR.Program) => TokenTree;
+export type Emitter = (
+  program: IR.Program,
+  addWarning: AddWarning,
+  compilationOptions: CompilationOptions
```

Comment:
Could be reasonable to combine `addWarning` and `compilationOptions` into a single "Context" object if they're often passed around together.

Reply:
True, this also came in mind.